rust_src/src/eval.rs

```diff
@@ -707,4 +707,50 @@ pub fn autoload(
 
 def_lisp_sym!(Qautoload, "autoload");
 
+/// Non-nil if OBJECT is a function.
+#[lisp_fn(name = "functionp", c_name = "functionp")]
+pub fn functionp_lisp(object: LispObject) -> bool {
+    FUNCTIONP(object.to_raw())
+}
+
+#[no_mangle]
+pub extern "C" fn FUNCTIONP(object: Lisp_Object) -> bool {
+    let mut obj = LispObject::from_raw(object);
+
+    if let Some(sym) = obj.as_symbol() {
+        if fboundp(sym) {
+            obj = sym.get_indirect_function();
+
+            if let Some(cons) = obj.as_cons() {
+                if cons.car().eq_raw(Qautoload) {
+                    // Autoloaded symbols are functions, except if they load
+                    // macros or keymaps.
+                    let mut it = obj.iter_tails_safe();
+                    for _ in 0..4 {
```

Comment:
This 4 is not obvious to me at all. Still, it's consistent with the C.

Reply:
Yeah, I agree. Definitely a wart to resolve.